crates/core/src/sql/compiler.rs

```diff
@@ -71,143 +71,112 @@ fn check_cmp_expr(table: &From, expr: &ColumnOp) -> Result<(), PlanError> {
 }
 
 /// Compiles a `WHERE ...` clause
-fn compile_where(q: QueryExpr, table: &From, filter: Selection) -> Result<QueryExpr, PlanError> {
+fn compile_where(mut q: QueryExpr, table: &From, filter: Selection) -> Result<QueryExpr, PlanError> {
     check_cmp_expr(table, &filter.clause)?;
     let schema = &table.root;
-    let predicate = filter.clause;
-    Ok(match extract_search_argument(table, predicate) {
-        (Some(SearchArgument::Eq { pos, value }), None) => q.with_index_scan(pos, value, schema.into()),
-        (Some(SearchArgument::Eq { pos, value }), Some(op)) => {
-            q.with_index_scan(pos, value, schema.into()).with_select(op)
+    for op in filter.clause.to_vec() {
+        match is_sargable(table, &op) {
+            Some(IndexArgument::Eq { col_id, value }) => {
+                q = q.with_index_eq(schema.into(), col_id, value);
+            }
+            Some(IndexArgument::LowerBound {
+                col_id,
+                value,
+                inclusive,
+            }) => {
+                q = q.with_index_lower_bound(schema.into(), col_id, value, inclusive);
+            }
+            Some(IndexArgument::UpperBound {
+                col_id,
+                value,
+                inclusive,
+            }) => {
+                q = q.with_index_upper_bound(schema.into(), col_id, value, inclusive);
+            }
+            None => {
+                q = q.with_select(op);
+            }
         }
-        (None, Some(op)) => q.with_select(op),
-        (None, None) => unreachable!(),
-    })
+    }
+    Ok(q)
 }
 
-// A SearchArgument is derived from an equality condition within a sargable
-// predicate.
-//
-// Sargable stands for Search ARGument ABLE.
-// A sargable predicate is one that can be answered using an index. In our case
-// it is a conjunction with at least one equality condition `<field> = <value>`.
-pub enum SearchArgument {
-    Eq { pos: u32, value: AlgebraicValue },
+// IndexArgument represents an equality or range predicate that can be answered
+// using an index.
+pub enum IndexArgument {
+    Eq {
+        col_id: u32,
```

Comment:
P.D: `col_id` elsewhere is not a unique id of the column but a position of it in the schema.